atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKvsAsyncFallbackMechanismsTests.java

```diff
@@ -153,16 +152,52 @@ public void testGetAsyncFallingBackToSynchronousOnInvalidAsyncKvs() {
         verify(keyValueService).get(TEST_TABLE, TIMESTAMP_BY_CELL);
     }
 
-    // TODO: Cleanup!! DO NOT MERGE
     @Test
-    public void testGetAsyncFallingBackToSynchronousOnSessionClosed() {
+    public void testGetAsyncFallingBackToSynchronousOnSessionClosedForExecuteAsync()
+            throws ExecutionException, InterruptedException {
+        CassandraKeyValueServiceConfig config = getConfigWithAsyncFactoryUsingClosedSession(true);
+
+        keyValueService = spy(CassandraKeyValueServiceImpl.createForTesting(config));
+        keyValueService.createTable(TEST_TABLE, AtlasDbConstants.GENERIC_TABLE_METADATA);
+
+        keyValueService.getAsync(TEST_TABLE, TIMESTAMP_BY_CELL).get();
+
+        verify(keyValueService).get(TEST_TABLE, TIMESTAMP_BY_CELL);
+    }
+
+    @Test
+    public void testGetAsyncFallingBackToSynchronousOnSessionClosedBeforeStatementPreparation() {
+        CassandraKeyValueServiceConfig config = getConfigWithAsyncFactoryUsingClosedSession(false);
+        keyValueService = spy(CassandraKeyValueServiceImpl.createForTesting(config));
+        keyValueService.createTable(TEST_TABLE, AtlasDbConstants.GENERIC_TABLE_METADATA);
+
+        keyValueService.getAsync(TEST_TABLE, TIMESTAMP_BY_CELL);
+
+        verify(keyValueService).get(TEST_TABLE, TIMESTAMP_BY_CELL);
+    }
+
+    private CassandraKeyValueServiceConfig getConfigWithAsyncFactoryUsingClosedSession(
+            boolean useSpyPreparedStatement) {
         CassandraKeyValueServiceConfig config = CASSANDRA_RESOURCE.getConfig();
         Cluster cluster = spy(new ClusterFactory(CASSANDRA_RESOURCE.getClusterBuilderWithProxy())
                 .constructCluster(CassandraClusterConfig.of(config), config.servers()));
         Session session = spy(cluster.connect());
 
         doReturn(session).when(cluster).connect();
 
+        if (useSpyPreparedStatement) {
+            // Need a real query that uses namespace and table that already exist, irrespective of what the test may
+            // create
+            PreparedStatement preparedStatement = spy(session.prepare("SELECT COUNT(*) FROM system.schema_columns;"));
+            BoundStatement boundStatement = spy(preparedStatement.bind());
+            // The prepared statement doesn't use any of the bindings that a normal query would use, so we ignore them.
+            doReturn(boundStatement).when(boundStatement).setBytes(any(), any());
+            doReturn(boundStatement).when(boundStatement).setLong(anyString(), anyLong());
+
+            doReturn(boundStatement).when(preparedStatement).bind();
```

Comment:
Discussed offline and noted in a comment above the hightlighted code - the query doesn't really match what `getAsync` expects, and `getAsync` tries to bind some parameters to the query that don't really apply. The mocking here just no-ops those, since I couldn't really find a working query that has those parameters.